azure_li_services/units/cleanup.py

```diff
@@ -32,3 +34,39 @@ def main():
             'azure-li-services'
         ]
     )
+
+    service_reports = Defaults.get_service_reports()
+
+    reboot_system = False
+    for report in service_reports:
+        if not report.get_state():
+            # in case a service has unknown or failed state we will
+            # not consider to reboot the machine
+            return
+        if report.get_reboot():
+            reboot_system = True
+
+    if reboot_system:
+        Command.run(
+            [
+                'kexec',
+                '--load', '/boot/vmlinuz',
+                '--initrd', '/boot/initrd',
+                '--command-line', get_boot_cmdline()
+            ]
+        )
+        Command.run(
+            ['kexec', '--exec']
+        )
+
```

Comment:
I have no idea what you mean ? we only reboot if _all_ services came back with a success result and if there is a request for reboot in at least one of the service results. In this situation you will not have any errors reported from the services just success, which I  consider the safe situation for a reboot. After reboot the systemd log messages from the services are gone, this is correct but not a problem in my eyes

Reply:
Let me explain better, sorry for the brevity. As I read through the code to handle reboots, it just felt like there were a number of reasons why a reboot would (or would not) happen. It would be nice to have a "blow by blow" log file stating what is happening (or not happening), and why. This sort of thing is typically easier to debug from than something like `systemctl` that just gives the final state.

But, as I see things, azure-li-services doesn't really have that sort of logging capability. All of it's logging is against `systemctl` services that give the end state, and no record of how the relevant code got there.

So this was more of a question: If we have that, this segment of code should log to it a lot more verbosely ... but again, I don't believe we do (can't see how we can and I would have missed it).

Reply:
That we do not have "formal" logging is dictated by the nature of the code. Some of the services have to run really early in the boot process when logging to syslog or access to /ar/log may not yet be available. Given the nature of the implementation its is going to be difficult to have a tempfile in memory that can be shared by all the services as each service is a separate process.

We can investigate in more depth how long we can delay the services to ensure /var/log is available and then log to a file. But this is a new feature request and should be handled separately. Given that we already had some timing issues with some of the services I am nt certain we can tweak the order.

Reply:
and services are designed to run in parallel as much as possible which I think is a concept we should keep for performance reasons. Thus having a serialized log file impacts performance.
I agree if logging is an issue please move the discussion to a new issue or we discuss by mail.

One word on logging though: I intentionally did not log formal processing as we delete all services and even the code for it once everything is done. From my point of view the services should provide detailed information on errors if there are any. The information here is meant for developers and should contain references to code and error information which is what you see when inspecting the systemctl status data. The reason why I'm after detailed error information is because we are always in an offline situation, meaning without access to the system. However if everything was successful there is probably no need to check on formal processing log information, especially not if all code is gone at that point in time.

So if the log information turns out to be an issue we need to discuss the "why" in another thread

Thanks